dissect/target/plugins/os/unix/log/journal.py

```diff
@@ -0,0 +1,393 @@
+
+    def entry_object_offsets(self) -> Iterator[int]:
+        """Read object entry arrays."""
+
+        offset = self.entry_array_offset
+
+        # Entry Array with next_entry_array_offset set to 0 is the last in the list
+        while offset != 0:
+            self.fh.seek(offset)
+
+            object = c_journal.ObjectHeader(self.fh)
+
+            if object.type == c_journal.ObjectType.OBJECT_ENTRY_ARRAY:
+                # After the object is checked, read again but with EntryArrayObject instead of ObjectHeader
+                self.fh.seek(offset)
+                entry_array_object = c_journal.EntryArrayObject(self.fh)
+
+                for entry_object_offset in entry_array_object.entry_object_offsets:
+                    # Check if the offset is not zero and points to nothing
+                    if entry_object_offset:
+                        yield entry_object_offset
+
+                offset = entry_array_object.next_entry_array_offset
+
+    def decode_value(self, value: bytes) -> tuple[str, str]:
+        value = value.decode().strip()
+
+        # Strip leading underscores part of the field name
+        value = value.lstrip("_")
+
+        key, value = value.split("=", 1)
+        key = key.lower()
+
+        return key, value
+
+    def __iter__(self) -> Iterator[Instance]:
+        "Iterate over the entry objects to read payloads."
+
+        for offset in self.entry_object_offsets():
+            self.fh.seek(offset)
+            entry = c_journal.EntryObject(self.fh)
+
+            event = {}
+            event["ts"] = ts.from_unix_us(entry.realtime)
+
+            for item in entry.items:
+                try:
+                    self.fh.seek(item.object_offset)
+
+                    object = c_journal.ObjectHeader(self.fh)
+
+                    if object.type == c_journal.ObjectType.OBJECT_DATA:
+                        self.fh.seek(item.object_offset)
+
+                        data_object = c_journal.DataObject(self.fh)
+
+                        data = data_object.payload
+
+                        if not data:
+                            # If the payload is empty
+                            continue
+                        elif data_object.flags == c_journal.ObjectFlag.OBJECT_COMPRESSED_XZ:
+                            data = lzma.decompress(data)
+                        elif data_object.flags == c_journal.ObjectFlag.OBJECT_COMPRESSED_LZ4:
+                            data = lz4.decompress(data[8:])
+                        elif data_object.flags == c_journal.ObjectFlag.OBJECT_COMPRESSED_ZSTD:
+                            data = zstandard.decompress(data)
+
+                        key, value = self.decode_value(data)
+                        event[key] = value
+
+                except Exception as e:
+                    self.target.log.warning(
+                        "The data object in Journal file %s could not be parsed", self.fh.name, exc_info=e
```

Comment:
Thanks, this is very nice. Unfortunately self.fh.name won't always work, since fh can be a RunlistStream (for instance if you open a .vbox). So maybe see if fh has a name attr first?